File: C++/Trees/treap_algo.cpp
#include <iostream>
#include <cstdlib>

using namespace std;

struct Node {
    int key;
    int priority;
    Node* left;
    Node* right;
};
 // Function to create a new node with the given key
// and a randomly generated priority value
Node* createNode(int key, int priority) {
    Node* newNode = new Node();
    newNode->key = key;
    newNode->priority = priority;
    newNode->left = newNode->right = NULL;
    return newNode;
}
// Function to perform a right rotation on the given root node
Node* rightRotate(Node* root) {
    Node* newRoot = root->left;
    Node* temp = newRoot->right;
    newRoot->right = root;
    root->left = temp;
    return newRoot;
}
// Function to perform a left rotation on the given root node
Node* leftRotate(Node* root) {
    Node* newRoot = root->right;
    Node* temp = newRoot->left;
    newRoot->left = root;
    root->right = temp;
    return newRoot;
}
// Function to insert a new node with the given key into the tree
Node* insert(Node* root, int key, int priority) {
    if (root == NULL)
        return createNode(key, priority);

    if (key < root->key) {
        root->left = insert(root->left, key, priority);
        if (root->left->priority > root->priority)
            root = rightRotate(root);
    }
    else {
        root->right = insert(root->right, key, priority);
        if (root->right->priority > root->priority)
            root = leftRotate(root);
    }
    return root;
}
 // Function to delete the node with the given key from the tree
Node* deleteNode(Node* root, int key) {
    if (root == NULL)
        return root;

    if (key < root->key)
        root->left = deleteNode(root->left, key);
    else if (key > root->key)
        root->right = deleteNode(root->right, key);
    else {
        if (root->left == NULL) {
            Node* temp = root->right;
            delete root;
            root = temp;
        }
        else if (root->right == NULL) {
            Node* temp = root->left;
            delete root;
            root = temp;
        }
        else {
            if (root->left->priority > root->right->priority) {
                root = rightRotate(root);
                root->right = deleteNode(root->right, key);
            }
            else {
                root = leftRotate(root);
                root->left = deleteNode(root->left, key);
            }
        }
    }
    return root;
}
// Function to perform an inorder traversal of the tree
void inorderTraversal(Node* root) {
    if (root != NULL) {
        inorderTraversal(root->left);
        cout << "Key: " << root->key << " Priority: " << root->priority << endl;
        inorderTraversal(root->right);
    }
}

int main() {
    Node* root = NULL;
    int choice, key, priority;
//Main loop to interact with the user and perform operations
    while (true) {
        cout << "1. Insert  2. Delete  3. Print Inorder  4. Exit\n";
        cout << "Enter your choice: ";
        cin >> choice;

        switch (choice) {
            case 1:
                cout << "Enter the key to insert: ";
                cin >> key;
                cout << "Enter the priority: ";
                cin >> priority;
                root = insert(root, key, priority);
                break;

            case 2:
                cout << "Enter the key to delete: ";
                cin >> key;
                root = deleteNode(root, key);
                break;

            case 3:
                cout << "Inorder Traversal:" << endl;
                inorderTraversal(root);
                break;

           case 4:
                exit(0);

            default:
                cout << "Invalid choice. Please try again.\n";
                break;
        }
        cout << endl;
    }

    return 0;
}

File: Python/Trees/treap_algo.py
import random

class Node:
   #Function to create a new node with the given key
# and a randomly generated priority value
    def __init__(self, key, priority):
        self.key = key
        self.priority = priority
        self.left = None
        self.right = None
# Function to perform a right rotation on the given root node
def right_rotate(root):
    new_root = root.left
    temp = new_root.right
    new_root.right = root
    root.left = temp
    return new_root
# Function to perform a left rotation on the given root node
def left_rotate(root):
    new_root = root.right
    temp = new_root.left
    new_root.left = root
    root.right = temp
    return new_root
#Function to insert a new node with the given key into the tree
def insert(root, key, priority):
    if root is None:
        return Node(key, priority)

    if key < root.key:
        root.left = insert(root.left, key, priority)
        if root.left.priority > root.priority:
            root = right_rotate(root)
    else:
        root.right = insert(root.right, key, priority)
        if root.right.priority > root.priority:
            root = left_rotate(root)
    return root
# Function to delete the node with the given key from the tree
def delete_node(root, key):
    if root is None:
        return root

    if key < root.key:
        root.left = delete_node(root.left, key)
    elif key > root.key:
        root.right = delete_node(root.right, key)
    else:
        if root.left is None:
            temp = root.right
            root = None
            return temp
        elif root.right is None:
            temp = root.left
            root = None
            return temp

        if root.left.priority > root.right.priority:
            root = right_rotate(root)
            root.right = delete_node(root.right, key)
        else:
            root = left_rotate(root)
            root.left = delete_node(root.left, key)
    return root
# Function to perform an inorder traversal of the tree
def inorder_traversal(root):
    if root:
        inorder_traversal(root.left)
        print("Key:", root.key, "Priority:", root.priority)
        inorder_traversal(root.right)

root = None
# Main loop to interact with the user and perform operations
while True:
    print("1. Insert  2. Delete  3. Print Inorder  4. Exit")
    choice = int(input("Enter your choice: "))

    if choice == 1:
        key = int(input("Enter the key to insert: "))
        priority = int(input("Enter the priority: "))
        root = insert(root, key, priority)
    elif choice == 2:
        key = int(input("Enter the key to delete: "))
        root = delete_node(root, key)
    elif choice == 3:
        print("Inorder Traversal:")
        inorder_traversal(root)
    elif choice == 4:
        break
    else:
        print("Invalid choice. Please try again.")
    
    print()

